From e94c1fb907546faafb3509615943776d1ea37eb8 Mon Sep 17 00:00:00 2001
From: Daiki Ueno <ueno@gnu.org>
Date: Wed, 3 Sep 2025 17:10:21 +0900
Subject: [PATCH] rpc: Fix empty array attribute handling

When an empty array attribute is exchanged at the RPC level, the
client previously sent the number of elements (= 0) even if it's
empty, while the server doesn't expect it. This fixes the client to
not send it.

Signed-off-by: Daiki Ueno <ueno@gnu.org>
---
 p11-kit/rpc-message.c |  2 +-
 p11-kit/test-mock.c   | 12 ++++++++++++
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/p11-kit/rpc-message.c b/p11-kit/rpc-message.c
index 049417f..5eaea61 100644
--- a/p11-kit/rpc-message.c
+++ b/p11-kit/rpc-message.c
@@ -266,7 +266,7 @@ p11_rpc_message_write_attribute_buffer_array (p11_rpc_message *msg,
 		/* And the attribute buffer length */
 		p11_rpc_buffer_add_uint32 (msg->output, attr->pValue ? attr->ulValueLen : 0);
 
-		if (IS_ATTRIBUTE_ARRAY (attr))
+		if (attr->pValue && IS_ATTRIBUTE_ARRAY (attr))
 			p11_rpc_message_write_attribute_buffer_array (
 				msg, attr->pValue,
 				attr->ulValueLen / sizeof (CK_ATTRIBUTE));
diff --git a/p11-kit/test-mock.c b/p11-kit/test-mock.c
index b117b92..f174015 100644
--- a/p11-kit/test-mock.c
+++ b/p11-kit/test-mock.c
@@ -624,6 +624,12 @@ test_get_wrap_template (void)
 		{ CKA_WRAP_TEMPLATE, temp, sizeof (temp) },
 	};
 	CK_ULONG n_attrs = sizeof (attrs) / sizeof (attrs[0]);
+	CK_OBJECT_CLASS klass = -1ul;
+	CK_ATTRIBUTE attrs_empty_template[] = {
+		{ CKA_WRAP_TEMPLATE, NULL, 0 },
+		{ CKA_UNWRAP_TEMPLATE, NULL, 0 },
+	};
+	CK_ULONG n_attrs_empty_template = sizeof(attrs_empty_template) / sizeof(attrs_empty_template[0]);
 
 	module = setup_mock_module (&session);
 
@@ -664,6 +670,12 @@ test_get_wrap_template (void)
 	assert (verify == CK_TRUE);
 	assert (encrypt == CK_TRUE);
 
+	rv = (module->C_GetAttributeValue) (session, MOCK_PUBLIC_KEY_CAPITALIZE, attrs_empty_template, n_attrs_empty_template);
+	assert (rv == CKR_ATTRIBUTE_TYPE_INVALID);
+	assert_num_eq (attrs_empty_template[0].type, CKA_WRAP_TEMPLATE);
+	assert_ptr_eq (attrs_empty_template[0].pValue, NULL);
+	assert_num_eq (attrs_empty_template[0].ulValueLen, (CK_ULONG)-1);
+
 	teardown_mock_module (module);
 }
 
-- 
2.50.1

